File: libgomp/testsuite/libgomp.fortran/target-map-iterators-3.f90
! { dg-do run }

! Test transfer of dynamically-allocated arrays to target using map
! iterators, with multiple iterators and function calls in the iterator
! expression.

program test
  implicit none

  integer, parameter :: DIM1 = 16
  integer, parameter :: DIM2 = 4

  type :: array_ptr
    integer, pointer :: arr(:)
  end type

  type (array_ptr) :: x(DIM1), y(DIM1)
  integer :: expected, sum, i, j, k

  expected = mkarrays ()

  !$omp target map(iterator(i=0:DIM1/4-1, j=0:3), to: x(f (i, j))%arr(:)) &
  !$omp        map(iterator(k=1:DIM1), to: y(k)%arr(:)) &
  !$omp        map(from: sum)
    sum = 0
    do i = 1, DIM1
      do j = 1, DIM2
	sum = sum + x(i)%arr(j) * y(i)%arr(j)
      end do
    end do
  !$omp end target

  if (sum .ne. expected) stop 1
contains
  integer function mkarrays ()
    integer :: exp = 0

    do i = 1, DIM1
      allocate (x(i)%arr(DIM2))
      allocate (y(i)%arr(DIM2))
      do j = 1, DIM2
	x(i)%arr(j) = i * j
	y(i)%arr(j) = i + j
	exp = exp + x(i)%arr(j) * y(i)%arr(j)
      end do
    end do

    mkarrays = exp
  end function

  integer function f (i, j)
    integer, intent(in) :: i, j

    f = i * 4 + j + 1
  end function
end program

File: libgomp/testsuite/libgomp.c-c++-common/target-map-iterators-3.c
/* { dg-do run } */
/* { dg-require-effective-target offload_device_nonshared_as } */

/* Test transfer of dynamically-allocated arrays to target using map
   iterators, with multiple iterators and function calls in the iterator
   expression.  */

#include <stdlib.h>

#define DIM1 16
#define DIM2 15

int mkarrays (int *x[], int *y[])
{
  int expected = 0;

  for (int i = 0; i < DIM1; i++)
    {
      x[i] = (int *) malloc (DIM2 * sizeof (int));
      y[i] = (int *) malloc (sizeof (int));
      *y[i] = rand ();
      for (int j = 0; j < DIM2; j++)
	{
	  x[i][j] = rand ();
	  expected += x[i][j] * *y[i];
	}
    }

  return expected;
}

int f (int i, int j)
{
  return i * 4 + j;
}

int main (void)
{
  int *x[DIM1], *y[DIM1];
  int sum;

  int expected = mkarrays (x, y);

  #pragma omp target enter data map(to: x, y)
  #pragma omp target map(iterator(i=0:DIM1/4, j=0:4), to: x[f(i, j)][:DIM2]) \
		     map(iterator(i=0:DIM1), to: y[i][:1]) \
		     map(from: sum)
    {
      sum = 0;
      for (int i = 0; i < DIM1; i++)
	for (int j = 0; j < DIM2; j++)
	  sum += x[i][j] * y[i][0];
    }

  return sum - expected;
}
